cmds/core/timeout/main.go

```diff
@@ -0,0 +1,81 @@
+// Copyright 2012-2022 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Run a command and kill it if it runs more than a specified duration
+//
+// Synopsis:
+//	timeout [-t duration-string] command [args...]
+//
+// Description:
+//	timeout will run the command until it succeeds or too much time has passed.
+//	The default timeout is 30s.
+//	If no args are given, it will print a usage error.
+//
+// Example:
+//	$ timeout echo hi
+//	hi
+//	$
+//	$./timeout -t 5s bash -c 'sleep 40'
+//	$ 2022/03/31 14:47:32 signal: killed
+//	$./timeout  -t 5s bash -c 'sleep 40'
+//	$ 2022/03/31 14:47:40 signal: killed
+//	$./timeout  -t 5s bash -c 'sleep 1'
+//	$
+
+//go:build !test
+// +build !test
+
+package main
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"log"
+	"os"
+	"os/exec"
+	"time"
+)
+
+type cmd struct {
+	args         []string
+	timeout      time.Duration
+	in, out, err *os.File
+	exitcode     int
+}
+
+var (
+	timeout        = flag.Duration("t", 30*time.Second, "Timeout for command")
+	errNoArgs      = errors.New("Need at least a command to run")
+	errBadDuration = errors.New("timeout must be a duration")
+)
+
+func main() {
+	flag.Parse()
+	c := &cmd{args: flag.Args(), in: os.Stdin, out: os.Stdout, err: os.Stderr, timeout: *timeout}
+	if err := c.run(); err != nil {
+		log.Printf("timeout(%v):%v", *timeout, err)
+		os.Exit(c.exitcode)
+	}
+}
+
+func (c *cmd) run() error {
+	if len(c.args) == 0 {
+		c.exitcode = 1
+		return errNoArgs
+	}
+	ctx := context.Background()
+	ctx, cancel := context.WithTimeout(ctx, c.timeout)
+	defer cancel()
+	proc := exec.CommandContext(ctx, c.args[0], c.args[1:]...)
+	proc.Stdin, proc.Stdout, proc.Stderr = c.in, c.out, c.err
+	if err := proc.Run(); err != nil {
+		c.exitcode = 1
+		if execerr, ok := err.(*exec.ExitError); ok {
```

Comment:
https://go.dev/blog/go1.13-errors - errors.As might be helpful here, it's a bit cleaner than the type conversion

var e *exec.ExitError
if errors.As(err, &e) {
  c.exitcode = e.ExitCode()
}

Reply:
Done. My fingers are so used to typing the old form that I forgot :-)